### ISSUE TITLE
transit-react-basic example transaction declares authority but does not have permissions with scatter on local blockchain

### ISSUE THREAD

**Issue by NedelescuVlad** [eosnewyork/eos-transit#5]
Hi, I'm running the transit-react-basic example and have connected to Scatter Desktop successfully, having set up scatter to connect to my local test EOS blockchain node. After logging in via scatter on one of my test accounts (alice2) I am trying to transfer 1 SYS to another test account, but encounter this error:

"transaction declares authority '{"actor":"alice2","permission":"active"}', but does not have signatures for it."

Here is how I initialise the wallet context:
```
const walContext = initDefaultAccessContext({
	appName,
        network: {
		host: 'localhost',
                port: 8888,
                protocol: 'http',
                chainId: "cf057bbfb72640471fd910bcb67639c22df9f92470936cddc1ade0e2f2e7dc4f", // Localhost
	},
	walletProviders: [ scatter() ]
});
```

Here is what my dev console outputs:
![image](https://user-images.githubusercontent.com/14984176/53801908-59462b00-3f38-11e9-9321-cba61a022978.png)

And the request for get_required_keys that comes back with a 401 unauthorized:
![image](https://user-images.githubusercontent.com/14984176/53801948-7549cc80-3f38-11e9-877c-10d970b06b2b.png)

In the last screenshot we can see that `available_keys` is sent as an empty array. Shouldn't this contain the public key for alice2 (the user that I'm logged in as)?
I suspect that could be the reason but I don't know how to fix it. Is it a configuration error in my scatter (I seem to have imported alice2's private key correctly though and linked that key to the alice2 account)? Or is transit or scatterjs failing to retrieve the public key for alice2 from my scatter? 

Any help or advice is greatly appreciated.

**Comment by warrick-eosny** [comment]
Hi Vlad, this looks like a CORS issue on your local EOSIO configuration. 

Check that you have the following in your config file

access-control-allow-origin = *

By default the EOSIO software does not allow cross domain requests. He's another question with something similar: https://eosio.stackexchange.com/questions/457/eosjs-http-cors-error 

Let me know if that helps.

**Comment by NedelescuVlad** [comment]
Thanks for the quick reply Warrick. I'm already running nodeos with `access-control-allow-origin = *`. I'm digging deeper into eosjs right now as this doesn't really seem to be an issue with Transit.

**Comment by warrick-eosny** [comment]
Hi Vlad,

I would focus on that 401 you're getting from get_required_keys - that is what's causing all the problems. 

I would do the following:
1. Run the EOSIO servers. 
2. Use a tool like postman or fiddle to make that POST request to to get_required_keys  on the local URL and debug outside of eosjs.

Once you can make the basic http request then come back to eosjs and I suspect it'll all work as expected. 

Thanks
Warrick

**Comment by NedelescuVlad** [comment]
Thanks Warrick,

 I've made the request via Postman and am getting the same unauthorized error. I'm going to run the transit-react-basic example on the Penn network for now and check that it's fine; I'll be switching to Jungle network afterwards, deploy my eosio code to Jungle and test my integration there, as I expect there's something about my local setup that's causing the issue.

I'm closing this issue as I don't think it's relevant to Transit after all. Thank you for your time and advice. If I find the fix to the 401 error I'll try to remember and update this thread.

**Comment by NedelescuVlad** [comment]
I was able to reproduce with Penn network. I'll post an issue separately detailing steps to reproduce.

**Comment by warrick-eosny** [comment]
Interesting, ok, thanks. Will check it out.

**Comment by NedelescuVlad** [comment]
How to reproduce:
1. Clone latest master branch of eos-transit.
2. `yarn install` and `yarn bootstrap` from the project root
3. Configure scatter according to the README in the `examples/transit-react-basic` directory
4. `yarn start` from `examples/transit-react-basic`
5. Visit localhost in Chrome; this should give a regeneratorRuntime not found error. Skip step 6 if no such error occurs
6. To fix regeneratorRuntime error: `yarn run eject` from transit-react-basic dir. Followed by `yarn add babel-polyfill` and editing `webpack.config.dev.js` and `webpack.config.prod.js` as such: Add `"babel-polyfill`" under `entry` array, before `require.resolve('./polyfills')`
7. `yarn run start` and visit localhost; should work fine now.
8. Login with Scatter Desktop. Try to issue transaction to mary12345123. Transaction should be denied with "transaction declares authority '{"actor":"eostransitio","permission":"active"}', but does not have signatures for it."

**Comment by NedelescuVlad** [comment]
Thanks Warrick, this is not very urgent.

**Comment by warrick-eosny** [comment]
Strange I just tried this and it worked for me. 
1. Can you please confirm that you experience the same problem @ this URL https://www.demo.eostransit.io/
2. Please confirm that which account your signing the txn with in Scatter (I'm assuming eostransitio) but just confirming. 
3. Please confirm your OS and Version of Scatter

**Comment by NedelescuVlad** [comment]
1. Just tested on the demo website and it works perfectly.
2. Yes, it is eostransitio. I don't get to see the "Allow transaction" screen in scatter though as I do in the demo site, as it doesn't reach that point.
3. I'm on a Windows 10 machine. Scatter version is 10.1.2

**Comment by warrick-eosny** [comment]
That is very strange. mm .. I'll get back to you in a bit, checking a few things.

**Comment by NedelescuVlad** [comment]
My guess is that the 401 happens because nothing is provided in `available_keys` when running locally. I've looked into the eos repo on github and was able to confirm this. Here is the code that is hit in order of how the request is processed: 

https://github.com/EOSIO/eos/blob/3fddb727b8f3615917707281dfd3dd3cc5d3d66d/plugins/chain_plugin/chain_plugin.cpp#L1862
https://github.com/EOSIO/eos/blob/3fddb727b8f3615917707281dfd3dd3cc5d3d66d/libraries/chain/authorization_manager.cpp#L586 . `candidate_keys` here comes up as empty so the checker doesn't mark the tx as valid 
https://github.com/EOSIO/eos/blob/3fddb727b8f3615917707281dfd3dd3cc5d3d66d/plugins/http_plugin/http_plugin.cpp#L569

So imo the problem boils down to why `available_keys` is empty. Perhaps the reason for which `available_keys` comes up as empty on localhost is because scatter or scatterjs refuses to communicate with http websites on some of its operations, as a security measure?

**Comment by NedelescuVlad** [comment]
^ Tried https but nothing changed. I'll try using scatterjs directly.

**Comment by warrick-eosny** [comment]
When you run Scatter, Click the Cog in the top right corner. Then scroll to the bottom of the page and click "Open Console". Are there any errors listed in the Console?

**Comment by NedelescuVlad** [comment]
That's a great tip! Unfortunately no errors in the console. 

I have however managed to get things working with scatterjs and eosjs 16.0.9 (my code is here: https://codesandbox.io/s/kkqv64r753). Made a transfer to mary12345123 account from eostransitio account.

My next investigation would be into eosjs@beta3, after having found someone have the same issue on scatterjs repo: https://github.com/GetScatter/scatter-js/issues/92#issuecomment-460451898 (same error, same eosjs as transit). 

I think I'll try changing the transit scatter plugin to use eosjs 16.0.9 tomorrow (quite late in the UK right now) as I'd really like to use Transit . I'll let you know how it goes.

**Comment by warrick-eosny** [comment]
I am using eosjs@beta3 so this would be a good test. 

One way of the other we will get you working. I'm wrapping up some changes today but if you don't find anything obvious today then let's plan to do a google hangout tomorrow and we'll debug it together. There's something obvious that I'm missing here.

**Comment by NedelescuVlad** [comment]
That would be great. I'd be more than happy for a google hangouts tomorrow.

**Comment by NedelescuVlad** [comment]
I don't think a Google Hangouts is necessary atm. I was able to narrow down the issue to 
`wallet.provider.signatureProvider.getAvailableKeys().length == 0`, which means that something is amiss when calling the SignatureProvider constructor of eosjs@20.0.x (might be called without any private keys as constr arguments somehow). I'll look more into this when I have more time but at the moment I'm focusing on other aspects of my project. I'll keep you posted when I have updates.

**Comment by warrick-eosny** [comment]
I just had some else with a similar problem and it turned out he was using eosjs in his project and not eosjsv2 ... can you please confirm that the version being included in your build is eosjsv2 .. eg. eosjs@beta3 or higher

**Comment by warrick-eosny** [comment]
ok, so I finally figured this out ... and well the answer is pretty stupid. 

When I did a npm list ... this is what I was seeing in the case where I could replicate the problem. Notice the scatterjs-core@2.3.8 at the bottom. 
```
 grep scatter npm-list.txt
├─┬ eos-transit-scatter-provider@3.1.0
│ ├─┬ scatterjs-core@2.7.16
│ └── scatterjs-plugin-eosjs2@1.5.0
├─┬ scatterjs-core@2.3.8
```

You'll see that the latest example package I just committed has scatterjs-core@2.3.8 removed from it's dependencies. No idea what it was doing there, but it was conflicting with the newer scatterjs-core@2.7.16 package. 

So remove scatterjs-core@2.3.8 and then run npm list to make sure it's not there ... that'll take care of the problem.
